spec/Candid.md

```diff
@@ -864,142 +864,213 @@ service { <methtype>;* } <: service { <methtype'>;* }
-------------------------------------------------------------------------------
-record { <fieldtype>;* } <: record { <nat> : opt <datatype'>; <fieldtype'>;* }
-  ~> \x.{f x with <nat> = null}
+```
+<v> :? <t> ~> <v'>
+----------------------------------------------------------------------------------
+variant { <nat> = <v> } :? variant { <nat> = <t>; _;* } ~> variant { <nat> = <v'> }
 ```
 
 
-#### Variants
+#### References
+
+Function and services references decode unconditoinally
 
+```
+-------------------------------------------------------
+func <text>.<id> :? func <functype> ~> func <text>.<id>
 ```
 
--------------------------------------------------
-variant { } <: variant { <fieldtype'>;* } ~> \x.x
+```
+-------------------------------------------------------
+service <text> :? service <actortype> ~> service <text>
+```
 
-<datatype> <: <datatype'> ~> f1
-variant { <fieldtype>;* } <: variant { <fieldtype'>;* } ~> f2
-------------------------------------------------------------------------------------------------
-variant { <nat> : <datatype>; <fieldtype>;* } <: variant { <nat> : <datatype'>; <fieldtype'>;* }
-  ~> \x.case x of <nat> y => <nat> (f1 y) | _ => f2 x
+```
+-------------------------------------------------
+principal <text> :? principal ~> principal <text>
 ```
 
+#### Tuple types
 
-#### Functions
+Whole argument and result sequences are decoded with the same rules are tuple-like records. In particular, extra arguments are ignored, and optional parameters read as as `null` if the argument is missing or fails to decode:
 
 ```
-record { N1':<datatype1'>;* } <: record { N1:<datatype1>;* } ~> f1
-record { N2:<datatype2>;* } <: record { N2':<datatype2'>;* } ~> f2
-------------------------------------------------------------------------------------------------------------------
-func ( <datatype1>,* ) -> ( <datatype2>,* ) <funcann>* <: func ( <datatype1'>,* ) -> ( <datatype2'>,* ) <funcann>*
-  ~> \x.\y.f2 (x (f1 y))
+record {<v>;*} :? record {<t>;*} ~> record {<v'>,*}
+---------------------------------------------------
+(<v>,*) :? (<t>,*) ~> (<v'>,*)
 ```
 
-#### Services
 
-```
+## Properties
 
--------------------------------------------------
-service { <methtype'>;* } <: service { } ~> \x.{}
+The relations above have certain properties. To express them, we need the relation `V : T`, expressing that `V` has inherently type `T`. Instead of defining this relation on its own, we take the first property below as its definition:
 
-<functype> <: <functype'> ~> f1
-service { <methtype>;* } <: service { <methtype'>;* } ~> f2
-------------------------------------------------------------------------------------------------
-service { <name> : <functype>; <methtype>;* } <: service { <name> : <functype'>; <methtype'>;* }
-  ~> \x.{f1 x; <name> = f2 x.<name>}
-```
+* Correctness and completeness of decoding:
+  ```
+  (v : T) ⟺ (∃ v'. v' :? T ~> v)
+  ```
 
+* Roundtripping:
+  ```
+  (v : T) ⟺ v :? T ~> v
```

Comment:
Hm, is that roundtripping? Is this assuming that serialisation is a no-op in this abstract model? Maybe this needs to be stated somewhere.

Reply:
Yes, this means that in the abstract model, roundtripping is a no-op. Note that `v : T` is the “precise” has-type relation (no extra fields, for example), so this is possible.

Where should I state this?

Reply:
Maybe the intro to the whole Decoding should somehow explain how this abstract model relates to actual serialisation/deserialisation.